lxd/storage/backend_lxd.go

```diff
@@ -2223,6 +2299,174 @@ func (b *lxdBackend) CreateInstanceFromMigration(inst instance.Instance, conn io
+	if err != nil {
+		return err
+	}
+
+	if volExists {
+		return fmt.Errorf("Volume already exists on storage but not in database")
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Validate config and create database entry for new storage volume if not refreshing.
+	// Strip unsupported config keys (in case the export was made from a different type of storage pool).
+	err = VolumeDBCreate(b, inst.Project().Name, inst.Name(), args.Description, volType, false, vol.Config(), inst.CreationDate(), time.Time{}, contentType, false, true)
+	if err != nil {
+		return err
+	}
+
+	revert.Add(func() { _ = VolumeDBDelete(b, inst.Project().Name, inst.Name(), volType) })
+
+	// Generate the effective root device volume for instance.
+	err = b.applyInstanceRootDiskOverrides(inst, &vol)
+	if err != nil {
+		return err
+	}
+
+	// Override args.Name and args.Config to ensure volume is created based on instance.
+	args.Config = vol.Config()
+	args.Name = inst.Name()
+
+	if slices.Contains(args.ConversionOptions, "format") {
+		// When conversion option "format" is enabled, we need to upload the image to a temporary location
+		// before converting it into the desired format. The conversion cannot be done in-place, therefore
+		// the image has to be saved in an intermediate location.
+		conversionID := fmt.Sprintf("conversion_%s_%s", inst.Project().Name, inst.Name())
+		imgUploadPath := filepath.Join(shared.VarPath("backups"), conversionID)
+
+		var wrapper *ioprogress.ProgressTracker
+		if args.TrackProgress {
+			wrapper = migration.ProgressTracker(op, "block_progress", vol.Name())
+		}
+
+		// Setup progress tracker.
+		fromPipe := io.ReadCloser(conn)
+		if wrapper != nil {
+			fromPipe = &ioprogress.ProgressReader{
+				ReadCloser: fromPipe,
+				Tracker:    wrapper,
+			}
+		}
+
+		// Create new file in backups directory.
+		toFile, err := os.OpenFile(imgUploadPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0)
+		if err != nil {
+			return fmt.Errorf("Error opening file for writing %q: %w", imgUploadPath, err)
+		}
+
+		// Ensure temporary image in backups directory is removed regardless of the conversion success.
+		defer func() {
+			_ = toFile.Close()
+			_ = os.Remove(imgUploadPath)
+		}()
+
+		// Receive image into the recently created file.
+		_, err = io.Copy(toFile, fromPipe)
+		if err != nil {
+			return fmt.Errorf("Error copying from conversion connection to %q: %w", imgUploadPath, err)
+		}
+
+		// Create a volume and provide a volume filler that converts a received image from backups
+		// directory into an instance.
+		volCopy := drivers.NewVolumeCopy(vol)
+		volFiller := drivers.VolumeFiller{
+			Fill: b.imageConversionFiller(imgUploadPath),
+		}
+
+		err = b.driver.CreateVolume(volCopy.Volume, &volFiller, op)
+		if err != nil {
+			return err
+		}
+
+		revert.Add(func() { _ = b.driver.DeleteVolume(volCopy.Volume, op) })
+	} else {
+		// If migration header supplies a volume size, then use that as block volume size instead of pool default.
```

Comment:
It'd be good to have a comment here explaining why this else code path is required?

The comment below mentions container filesystem, but why would this be being sent to this end point, is it because lxd-migrate will always use the new conversion code path if its available?

Reply:
Yes, exactly. I've imagined that the conversion API would replace the existing migration to have everything handled in one place (with regards to `lxd-migrate` tool). This means that it has to also support the migration of containers.